test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java

```diff
@@ -142,19 +172,200 @@ public void testPstmtInvalidation() throws Throwable
+        long statementsInMemory = numberOfStatementsInMemory();
+        long statementsOnDisk = numberOfStatementsOnDisk();
+        assertEquals(statementsOnDisk, statementsInMemory);
+
+        // Drop prepared statements from cache only and ensure the cache empties out.
+        QueryProcessor.clearPreparedStatements(true);
+        assertEquals(0, numberOfStatementsInMemory());
+
+        // Load prepared statements and ensure the cache size matches max
+        QueryProcessor.instance.preloadPreparedStatements(PRELOAD_PAGE_SIZE);
+
+        long statementsInMemoryAfterLoading = numberOfStatementsInMemory();
+        // Ensure size of cache matches statements that were on disk before preload
+        assertEquals("Statements prepared - evicted (expected) does not match statements in memory (actual)",
+                     statementsOnDisk, statementsInMemoryAfterLoading);
+
+        // Number of statements on disk shold match memory
+        assertEquals(statementsInMemoryAfterLoading, numberOfStatementsOnDisk());
+
+        // Ensure only executed the expected amount of pages.
+        int expectedPageInvocations = (int) Math.ceil(statementsInMemoryAfterLoading / (double) PRELOAD_PAGE_SIZE);
+        assertEquals(expectedPageInvocations, pageInvocations.get());
+    }
+
+    @Test
+    public void testPreloadPreparedStatementsUntilCacheFull()
+    {
+        ClientState clientState = ClientState.forInternalCalls();
+        createTable("CREATE TABLE %s (key int primary key, val int)");
+
+        // Fill up and clear the prepared statement cache several times to load up the system.prepared_statements table.
+        // This simulates a 'leak' of prepared statements akin to CASSANDRA-19703 as the system.prepared_statements
+        // table is able to grow to a larger size than the in memory prepared statement cache.  In such a case we
+        // should detect a possible leak and defer paging indefinitely by returning early in preloadPreparedStatements.
+        int fillIterations = 5;
+        int cnt = 1;
+        int statementsLoadedBeforeFull = -1;
+        for (int fillIteration = 0; fillIteration < fillIterations; fillIteration++)
+        {
+            while (!QueryProcessor.isPreparedStatementCacheNearlyFull())
+            {
+                prepareStatement("INSERT INTO %s (key, val) VALUES (?, ?) USING TIMESTAMP " + cnt, clientState);
+               cnt++;
+            }
+            if (statementsLoadedBeforeFull == -1) {
+                statementsLoadedBeforeFull = cnt;
+            }
+            QueryProcessor.clearPreparedStatements(true);
+        }
+
+        // Ensure the page size we use is less than the number of statments it took to load the cache.
+        int pageSize = statementsLoadedBeforeFull / 10;
+        // We should only expect to load how many statements we were able to load before filling the cache
+        // + a buffer of a page size + another possible page.
```

Comment:
we allow a buffer of 2 pages for:

one page to hit the boundary where we detect we might have leaked + 
additional page that we are allowed to read before exiting early.

Reply:
I would trash that and just loop and break when you go over the limit. it's just 2 local variable reads on startup and you drop unnecessary complexity of the code.